pymc/distributions/mixture.py

```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
+            )
 
-        if all_discrete(comp_dists):
-            default_dtype = _conversion_map[aesara.config.floatX]
+        w = at.as_tensor_variable(w)
+
+        # ShapeWarning does not make sense here
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", ShapeWarning)
+            return super().dist([w, *comp_dists], **kwargs)
+
+    @classmethod
+    def rv_op(cls, weights, *components, size=None, rngs=None):
+        # Update rngs if provided
+        if rngs is not None:
+            *components_rngs, choices_rng = rngs
+            new_components = []
+            assert len(components) == len(components_rngs)
+            for component, component_rng in zip(components, components_rngs):
+                component_node = component.owner
+                old_rng, *inputs = component_node.inputs
+                new_components.append(
+                    component_node.op.make_node(component_rng, *inputs).default_output()
+                )
+            components = new_components
         else:
-            default_dtype = aesara.config.floatX
-
-            try:
-                self.mean = (w * self._comp_means()).sum(axis=-1)
-
-                if "mean" not in defaults:
-                    defaults.append("mean")
-            except AttributeError:
-                pass
-        dtype = kwargs.pop("dtype", default_dtype)
-
-        try:
-            if isinstance(comp_dists, Distribution):
-                comp_mode_logps = comp_dists.logp(comp_dists.mode)
-            else:
-                comp_mode_logps = at.stack([cd.logp(cd.mode) for cd in comp_dists])
+            # Create new rng for the choices internal RV
+            choices_rng = aesara.shared(np.random.default_rng())
 
-            mode_idx = at.argmax(at.log(w) + comp_mode_logps, axis=-1)
-            self.mode = self._comp_modes()[mode_idx]
+        # Create a OpFromGraph that encapsulates the random generating process
+        weights_type = weights.type()
+        components_type = [component.type() for component in components]
+        rng_type = choices_rng.type()
 
-            if "mode" not in defaults:
-                defaults.append("mode")
-        except (AttributeError, ValueError, IndexError):
-            pass
-
-        super().__init__(shape, dtype, defaults=defaults, *args, **kwargs)
-
-    @property
-    def comp_dists(self):
-        return self._comp_dists
-
-    @comp_dists.setter
-    def comp_dists(self, comp_dists):
-        self._comp_dists = comp_dists
-        if isinstance(comp_dists, Distribution):
-            self._comp_dist_shapes = to_tuple(comp_dists.shape)
-            self._broadcast_shape = self._comp_dist_shapes
-            self.comp_is_distribution = True
+        stacked_components_ = at.stack(components_type, axis=-1)
+        if weights.ndim < stacked_components_.ndim:
+            weights_ = at.shape_padaxis(weights_type, axis=weights_type.ndim - 1)
```

Comment:
This is very likely completely wrong. I just trial and errored to get the tests to pass. From what @Sayam753 said, it should probably be based on the support dimensionality of the components, not on the dimensionality of the weights